tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
         }
         return false;
     }
-    
-    /* Read at most "this.buff.length" bytes into "this.buff",
-       returning the number of bytes read. If the return result
-       is less than "this.buff.length", then EOF was read. */
-    private int fillBuffer() throws IOException {
+
+    /**
+     * Restore invariant V3 by reading bytes into {@link #buff} starting from
+     * <code>super.{@link #getFilePointer()}</code>.  Call this after changing
+     * {@link #lo}.
+     *
+     * <p>Since this method clobbers the contents of {@link #buff}, callers
+     * must be careful to ensure that there are no unflushed bytes (i.e.
+     * {@link #dirty} is <code>false</code>). See {@link #flushBuffer()}.
+     */
+    private void fillBuffer() throws IOException {
+        assert !dirty && diskPos == super.getFilePointer();
+
+        // Ensure that we are reading from the correct point in the underlying file.
+        if (diskPos != lo) {
+            super.seek(lo);
+            diskPos = lo;
+        }
+
+        // Read until the buffer is full or EOF.
         int cnt = 0;
         int rem = this.buff.length;
         while (rem > 0) {
-            int n = super.read(this.buff, cnt, rem); 
-            if (n < 0) break;
+            int n = super.read(this.buff, cnt, rem);
+            if (n < 0) {
+                break; // EOF
+            }
             cnt += n;
             rem -= n;
         }
-        this.hitEOF = (cnt < this.buff.length);
+
+        // Update "this.diskPos" due to super.read() calls.
         this.diskPos += cnt;
-        return cnt;
+
+        // There is a subtle case to consider as this method exits.  Suppose that:
+        // 1. We didn't fill the buffer because we hit EOF (i.e. length < lo + BuffSz).
+        // 2. The client calls seek() to jump forward, followed by write() to write a byte later in the buffer.
+        //
+        // Then the unset part of buff (represented by "*" below) can actually matter:
+        //
+        //                 length was 4 when fillBuffer() was called
+        //                   v
+        //     buff = [A B C D * * * * ...]
+        //            >>>>>>>>>>>>>> ^
+        //            seek(7)      write(0xFF)
+        //
+        // In that case, the file must be extended to contain the write, and the unset part of the buffer becomes part
+        // of the official file contents.  Fortunately the RandomAccessFile contract gives us an out: when the file is
+        // extended in this way, the unwritten portion can be filled with arbitrary data.  Therefore, we do not need to
+        // do anything to the unset parts of buff.
+        //
+        // It is tempting to offer a slightly stronger contract by zero-filling the unset part of "buff":
+        //
+        //     Arrays.fill(this.buff, cnt, BuffSz, (byte)0);
+        //
+        // But do not be tempted!  That call incurs a performance penalty WITHOUT OFFERING A STRONGER CONTRACT!  There
```

Comment:
This is great to call out this optimization (or trap, depending how you're looking at it) and the reasoning behind it.